### ISSUE TITLE
Vectorize/Optimize the COCORecall metric

### ISSUE THREAD

**Issue by LukeWood** [keras-team/keras-cv#17]
Lots of the for loops in the COCORecall update_state method could be reduced to `tf.einsum()` calls.  Additionally, we should be able to vectorize area computation in the iou function, along with some other operations in the overall computation of the metric.  tf.TensorArrays are quite slow, so we should look to remove those if possible.

**Comment by bhack** [comment]
> Additionally, we should be able to vectorize area computation in the iou function

What version do you want to implement? We have a vectorized IOU in:
https://github.com/tensorflow/models/blob/master/official/vision/detection/utils/box_utils.py#L643

It is quite the same of the old vectorized benchmarks in:
https://medium.com/@venuktan/vectorized-intersection-over-union-iou-in-numpy-and-tensor-flow-4fa16231b63d

**Comment by bhack** [comment]
Also with the scope of having a reusable component library for CV we have also in TF Addons:

https://github.com/tensorflow/addons/blob/master/tensorflow_addons/losses/giou_loss.py#L107-L128

**Comment by LukeWood** [comment]
Wow, thank you for the vectorized iou.  This will save me a lot of time?

**Comment by LukeWood** [comment]
> > Additionally, we should be able to vectorize area computation in the iou function
> 
> What version do you want to implement? We have a vectorized IOU in: https://github.com/tensorflow/models/blob/master/official/vision/detection/utils/box_utils.py#L643
> 
> It is quite the same of the old vectorized benchmarks in: https://medium.com/@venuktan/vectorized-intersection-over-union-iou-in-numpy-and-tensor-flow-4fa16231b63d

moved to https://github.com/tensorflow/models/blob/master/official/legacy/detection/utils/box_utils.py

**Comment by LukeWood** [comment]
Profile trace of COCORecall:

![image](https://user-images.githubusercontent.com/12191303/151064395-8f517648-569f-4a4d-8bc5-5f80ccbf7dc4.png)

Looks like TensorScatterAdds make up for 20% of compute costs.  To fix this, I can use a TensorArray in COCOBase of shape (num_categories*num_thresholds) then reshape it after stacking.  This will be much more efficient.  

Additionally, we should look into the cost of the sum operations.

**Comment by bhack** [comment]
Can you share the gist to reproduce this as we still don't have any CI infra for PR performance regression check?

I suppose one of the problem is that  `TensorSCattterAdd` and `Sum` are in the inner loop. 
Is `StridedSlice` also coming from  `_match_boxes` innner loop?

The top 3 occurrences in the inner loops are going to take `48,5%` of the device time.

**Comment by LukeWood** [comment]
> Can you share the gist to reproduce this as we still don't have any CI infra for PR performance regression check?
Unfortunately no.  This uses an internal profiling tool.

> I suppose one of the problem is that TensorSCattterAdd and Sum are in the inner loop.
Yeah, the sum operation I expect we can't improve.  The ScatterND add, I would bet we can improve this drastically.  I will be updating the results to TensorArrays, which should help a lot.

Re: StridedSlice and matching boxes:
I also expect this to take a lot of compute time, it may just be an inherent truth of the algorithm required to compute the metric.

**Comment by bhack** [comment]
I think we need to be on the same page to test this also the same (simulated/random?) input size. Can you share your benchmarking gist?.

Have you checked if you are ok with the constrains imposed by:
https://www.tensorflow.org/api_docs/python/tf/vectorized_map

**Comment by bhack** [comment]
> moved to https://github.com/tensorflow/models/blob/master/official/legacy/detection/utils/box_utils.py

The new box and iou ops are in: 
https://github.com/tensorflow/models/tree/master/official/vision/beta/ops

**Comment by LukeWood** [comment]
> Unfortunately no.  This uses an internal profiling tool. :(
